File: 2023/ts/07/part02.ts
import fs from "fs";
import path from "path";
import { range } from "../utils";

{
  //   const text = `32T3K 765
  // T55J5 684
  // KK677 28
  // KTJJT 220
  // QQQJA 483
  // `;

  const text = fs.readFileSync(path.join(__dirname, "./input.txt"), {
    encoding: "ascii",
  });

  type Card = {
    labels: number[];
    bid: number;
    type: number;
  };

  const detType = (cnts: number[]): Card["type"] => {
    if (cnts[0]! === 5) {
      return 6; // five of a kind
    }
    if (cnts[0]! === 4) {
      return 5; // four of a kind
    }
    if (cnts[0]! === 3) {
      if (cnts[1]! === 2) {
        return 4; // full house
      }
      return 3; // three of a kind
    }
    if (cnts[0]! === 2) {
      if (cnts[1]! === 2) {
        return 2; // two pair
      }
      return 1; // one pair
    }
    return 0; // high card
  };

  const pretend = (type: Card["type"], js: number): Card["type"] => {
    switch (js) {
      case 0:
        return type;
      case 1:
        switch (type) {
          case 5:
            return 6;
          case 3:
            return 5;
          case 2:
            return 4;
          case 1:
            return 3;
          case 0:
            return 1;
          default:
            return type;
        }
      case 2:
        switch (type) {
          case 4:
            return 6;
          case 2:
            return 5;
          case 1:
            return 3;
          default:
            return type;
        }
      case 3:
        switch (type) {
          case 4:
            return 6;
          case 3:
            return 5;
          default:
            return type;
        }
      case 4:
        switch (type) {
          case 5:
            return 6;
          default:
            return type;
        }
      case 5:
        return 6;
      default:
        throw new Error(`invalid card type: ${type}`);
    }
  };

  const cards = text
    .trim()
    .split("\n")
    .map((line) => {
      const [labelsS, bidS] = line.split(/\s+/) as [string, string];
      const labels = Array.from(labelsS).map((v) => {
        if (v === "T") {
          return 10;
        }
        if (v === "J") {
          return 1;
        }
        if (v === "Q") {
          return 12;
        }
        if (v === "K") {
          return 13;
        }
        if (v === "A") {
          return 14;
        }
        return Number(v);
      });
      const bid = Number(bidS);
      const cnter = range(15).map(() => 0);
      for (const label of labels) {
        ++cnter[label];
      }
      const cnts = Array.from(cnter.values()).sort((a, b) => b - a);
      return {
        labels,
        bid,
        type: pretend(detType(cnts), cnter[1]!),
      } satisfies Card;
    });

  const sum = cards
    .sort(
      (a, b) =>
        a.type - b.type ||
        a.labels[0]! - b.labels[0]! ||
        a.labels[1]! - b.labels[1]! ||
        a.labels[2]! - b.labels[2]! ||
        a.labels[3]! - b.labels[3]! ||
        a.labels[4]! - b.labels[4]!
    )
    .reduce((acc, v, i) => acc + v.bid * (i + 1), 0);

  console.log(sum); // 254083736
}

File: 2023/rust/day07/src/part02.rs
use std::{cmp::Ordering, fs};

#[derive(PartialEq, Eq, PartialOrd, Ord, Debug)]
enum Type {
    FiveOfAKind,
    FourOfAKind,
    FullHouse,
    ThreeOfAKind,
    TwoPair,
    OnePair,
    HighCard,
}

#[derive(PartialEq, Eq, Debug)]
struct Card {
    labels: Vec<u32>,
    bid: u64,
    typ: Type,
}

impl Card {
    pub fn new(labels: Vec<u32>, bid: u64) -> Card {
        let mut cnter = [0; 15];
        for &label in labels.iter() {
            cnter[label as usize] += 1;
        }
        let mut sorted_cnter = cnter.iter().collect::<Vec<_>>();
        sorted_cnter.sort_by(|a, b| b.cmp(a));
        let typ = match sorted_cnter[..] {
            [&5, ..] => Type::FiveOfAKind,
            [&4, ..] => Type::FourOfAKind,
            [&3, &2, ..] => Type::FullHouse,
            [&3, _, ..] => Type::ThreeOfAKind,
            [&2, &2, ..] => Type::TwoPair,
            [&2, &1, ..] => Type::OnePair,
            _ => Type::HighCard,
        };
        Card {
            labels,
            bid,
            typ: pretend(typ, cnter[1]),
        }
    }
}

impl PartialOrd for Card {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Card {
    fn cmp(&self, other: &Self) -> Ordering {
        match other.typ.cmp(&self.typ) {
            Ordering::Equal => self.labels.cmp(&other.labels),
            v => v,
        }
    }
}

fn pretend(typ: Type, js: u32) -> Type {
    match js {
        0 => typ,
        1 => match typ {
            Type::FourOfAKind => Type::FiveOfAKind,
            Type::ThreeOfAKind => Type::FourOfAKind,
            Type::TwoPair => Type::FullHouse,
            Type::OnePair => Type::ThreeOfAKind,
            Type::HighCard => Type::OnePair,
            _ => typ,
        },
        2 => match typ {
            Type::FullHouse => Type::FiveOfAKind,
            Type::TwoPair => Type::FourOfAKind,
            Type::OnePair => Type::ThreeOfAKind,
            _ => typ,
        },
        3 => match typ {
            Type::FullHouse => Type::FiveOfAKind,
            Type::ThreeOfAKind => Type::FourOfAKind,
            _ => typ,
        },
        4 => match typ {
            Type::FourOfAKind => Type::FiveOfAKind,
            _ => typ,
        },
        5 => Type::FiveOfAKind,
        _ => panic!("invalid card type: {typ:?}"),
    }
}

pub fn solve() {
    // let text = "32T3K 765
    // T55J5 684
    // KK677 28
    // KTJJT 220
    // QQQJA 483
    // ";
    let text = fs::read_to_string("input.txt").unwrap();

    let mut cards = text
        .trim()
        .split("\n")
        .map(|line| {
            let line = line.trim();
            let labels_bid_s = line.split(" ").collect::<Vec<_>>();
            let labels = labels_bid_s[0]
                .chars()
                .map(|c| match c {
                    'T' => 10,
                    'J' => 1,
                    'Q' => 12,
                    'K' => 13,
                    'A' => 14,
                    _ => c.to_digit(10).unwrap(),
                })
                .collect::<Vec<_>>();
            Card::new(labels, labels_bid_s[1].parse::<u64>().unwrap())
        })
        .collect::<Vec<_>>();

    cards.sort();

    let res = cards
        .into_iter()
        .enumerate()
        .fold(0, |acc, (i, v)| acc + v.bid * ((i as u64) + 1));

    println!("{res}"); // 254083736
}
